overlord/snapstate/aliasesv2.go

```diff
@@ -0,0 +1,383 @@
+			status = "auto"
+		}
+		newStates[alias] = &AliasState{Status: status, Target: target}
+	}
+	if auto {
+		// nothing else to do
+		return newStates, nil
+	}
+	// carry over the current manual ones
+	for alias, curState := range curStates {
+		if !curState.Manual() {
+			continue
+		}
+		if info.Apps[curState.Target] == nil {
+			// not an existing app
+			continue
+		}
+		newState := newStates[alias]
+		if newState == nil {
+			newStates[alias] = &AliasState{Status: "manual", Target: curState.Target}
+		} else {
+			// alias is both manually setup but has an overlapping auto-alias
+			newStates[alias] = &AliasState{Status: "overridden", Target: curState.Target, AutoTargetBak: newState.Target}
+		}
+	}
+	return newStates, nil
+}
+
+type AliasConflictError struct {
+	Snap      string
+	Alias     string
+	Reason    string
+	Conflicts map[string][]string
+}
+
+func (e *AliasConflictError) Error() string {
+	if len(e.Conflicts) != 0 {
+		errParts := []string{"cannot enable"}
+		first := true
+		for snapName, aliases := range e.Conflicts {
+			if !first {
+				errParts = append(errParts, "nor")
+			}
+			if len(aliases) == 1 {
+				errParts = append(errParts, fmt.Sprintf("alias %q", aliases[0]))
+			} else {
+				errParts = append(errParts, fmt.Sprintf("aliases %s", strutil.Quoted(aliases)))
+			}
+			if first {
+				errParts = append(errParts, fmt.Sprintf("for %q,", e.Snap))
+				first = false
+			}
+			errParts = append(errParts, fmt.Sprintf("already enabled for %q", snapName))
+		}
+		// TODO: add recommendation about what to do next
+		return strings.Join(errParts, " ")
+	}
+	return fmt.Sprintf("cannot enable alias %q for %q, %s", e.Alias, e.Snap, e.Reason)
+}
+
+func checkAgainstEnabledAliasStates(st *state.State, checkedSnap string, checker func(alias, otherSnap string) error) error {
+	var allAliasStates map[string]map[string]*AliasState
+	err := st.Get("aliases-v2", &allAliasStates)
+	if err == state.ErrNoState {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	for otherSnap, aliasStates := range allAliasStates {
+		if otherSnap == checkedSnap {
+			// skip
+			continue
+		}
+		for alias, aliasState := range aliasStates {
+			if aliasState.Enabled() {
+				if err := checker(alias, otherSnap); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// checkAliasStatesConflicts checks candStates for conflicts against other snaps' alias states returning conflicting snaps and aliases for alias conflicts.
+func checkAliasStatesConflicts(st *state.State, snapName string, candStates map[string]*AliasState) (conflicts map[string][]string, err error) {
+	var snapNames map[string]*json.RawMessage
+	err = st.Get("snaps", &snapNames)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+
+	enabled := make(map[string]bool, len(candStates))
+	for alias, candState := range candStates {
+		if candState.Enabled() {
+			enabled[alias] = true
+		}
+		// check against snap namespaces
+		for name := range snapNames {
```

Comment:
wouldn't it be better to check `snapNames[alias-up-to-first-dot]` directly instead of looping?